[web-animations] start time of animations created during a page rendering update should match that update's timeline time

graouts opened the pull request:
#### 483c60c629b8394400c134b686ec549c1cf2c0f5
<pre>
[web-animations] start time of animations created during a page rendering update should match that update&apos;s timeline time
<a href="https://bugs.webkit.org/show_bug.cgi?id=290993">https://bugs.webkit.org/show_bug.cgi?id=290993</a>
<a href="https://rdar.apple.com/138844796">rdar://138844796</a>

Reviewed by Dean Jackson.

WebKit differs from Chrome and Firefox when setting the start time of animations
created during a page rendering update.

In WebKit, we mark any animation that is pending as ready when we run the &quot;update
animations and send events&quot; [0] step, which means that all animations pending at
this point will share the same start time and will also have a current time of 0
when the ready promise is resolved.

In Chrome and Firefox, animations created during a page rendering update will have
their start time set to the timeline time of that same update, even though their
ready promise may be resolved at a later time, for instance during the next page
rendering update. This means that the current time of animations will likely not
be 0 when the ready promise is resolved.

This is particularly visible for CSS-originated animations which may be created
during a page rendering update as styles are recalculated and layout is updated
a few steps after animations are updated. As such, WebKit typically is one rendering
frame behind Chrome and Firefox when dealing with CSS Transitions and CSS Animations.

This behavior is not actually a spec violation since the start time and time where
the ready promise is resolved is left as implementation-dependent [1].

However, this is proving to be a web compatibility issue since we found that captions
may flicker on YouTube for instance, and this is due to the behavior described above.
In fact, for CSS Transitions, we introduced some ad-hoc current time computation code
in 226359@main and 226577@main to deal with some compatibility issues (which we will
remove in a followup to this patch).

In light of this further web compatibility issue, we change how we determine the start
time of animations to align with Chrome and Firefox.

When we cache the current time during a page rendering update, we enqueue a task in
`AnimationTimelinesController::cacheCurrentTime()` to run once page rendering update
is done. Any animation that is created in this interval is added to a list of pending
animations, and when our queued that task is performed we set the &quot;pending start time&quot;
of those pending animations.

Then during the next page rendering update, as `WebAnimation::tick()` is called, we
check whether we have a &quot;pending start time&quot; and use that time to be the animation&apos;s
committed start time.

This now means that any animation created during a page rendering update, including
`requestAnimationFrame()` callbacks and code responding to animation events, will
have its pending start time set to the timeline time used for that page rendering
update. During the _next_ page rendering update, those animations will be marked as
&quot;ready&quot; and their start time will be set to the recorded pending start time.

If an animation was created any time after the page rendering update steps had
completed, for instance in a `setTimeout()` call within a `requestAnimationFrame()`
callback, that animation will have its start time set to match the next page rendering
update and will be marked as ready during the following page rendering update.

This behavior is now tested with an additional subtest in the existing WPT test
`web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html`.
While that behavior isn&apos;t technically normative, as noted above, it seems like important
enough for Web compatibility that it should be part of the WPT suite.

The subtle changes in timing required some adaptation of WebKit-specific tests. For the
compositing tests, because animations and transitions start a frame sooner, we had to update
those tests to have an animation transform in the output and to make sure the animations were
longer to yield a stable animated first frame in the output.

[0] <a href="https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events">https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events</a>
[1] <a href="https://drafts.csswg.org/web-animations-1/#waiting-for-the-associated-effect">https://drafts.csswg.org/web-animations-1/#waiting-for-the-associated-effect</a>

* LayoutTests/compositing/backing/backface-visibility-flip-expected.txt:
* LayoutTests/compositing/backing/backface-visibility-flip.html:
* LayoutTests/compositing/backing/backing-store-attachment-empty-keyframe.html:
* LayoutTests/compositing/backing/transition-extent-expected.txt:
* LayoutTests/compositing/backing/transition-extent.html:
* LayoutTests/compositing/layer-creation/overlap-animation-clipping.html:
* LayoutTests/compositing/layer-creation/overlap-animation-container.html:
* LayoutTests/compositing/layer-creation/overlap-animation-expected.txt:
* LayoutTests/compositing/layer-creation/overlap-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html:
* LayoutTests/platform/ios/compositing/backing/transition-extent-expected.txt:
* LayoutTests/platform/mac/compositing/visible-rect/animated-from-none-expected.txt:
* LayoutTests/webanimations/css-transition-retargeting-to-same-value-upon-completion-with-timeout.html:
* LayoutTests/webanimations/custom-effect/custom-effect.html:
* LayoutTests/webanimations/offset-anchor-animation-yields-compositing.html:
* LayoutTests/webanimations/offset-distance-animation-yields-compositing.html:
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::animationTimingDidChange):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::detachFromDocument):
(WebCore::AnimationTimelinesController::addPendingAnimation):
(WebCore::AnimationTimelinesController::cacheCurrentTime):
(WebCore::AnimationTimelinesController::processPendingAnimations):
(WebCore::AnimationTimelinesController::maybeClearCachedCurrentTime): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::removeReplacedAnimations): Use the live current time instead
of the cached current time so that we can guarantee that &quot;remove&quot; events are dispatched
after &quot;finish&quot; events for the same animation.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::runPendingPlayTask): use the &quot;pending start time&quot; if set.
(WebCore::WebAnimation::runPendingPauseTask): use the &quot;pending start time&quot; if set.
(WebCore::WebAnimation::tick):
(WebCore::WebAnimation::maybeMarkAsReady): move &quot;ready&quot; logic into a dedicated method.
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setPendingStartTime):

Canonical link: <a href="https://commits.webkit.org/294049@main">https://commits.webkit.org/294049@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c58aba7e104ecff0250df942a58bcb44691fbb5d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100705 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20357 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10656 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51293 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102746 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28831 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/105842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/33716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103712 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/15866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/90960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/105842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/15675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/8968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/50669 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85576 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/9043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/108197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27823 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/108197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28186 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/87162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/108197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/29884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21813 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16377 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27758 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/33011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27569 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30887 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29127 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/DocumentTimeline.cpp:
```diff
@@ -299,10 +299,8 @@ void DocumentTimeline::removeReplacedAnimations()
-                    ASSERT(currentTime->time());
-                    return documentTimeline->convertTimelineTimeToOriginRelativeTime(*currentTime->time());
-                }
+                auto currentTime = MonotonicTime::now().secondsSinceEpoch();
```
Shouldn't this use some snapshotted time, like the time for the current event loop update?

graouts replied:
Here's what I put in the changelog about this change:

> Use the live current time instead of the cached current time so that we can guarantee that "remove" events are dispatched after "finish" events for the same animation.

This scheduling time is not exposed to the Web and is used within `compareStyleOriginatedAnimationEvents()` for event sorting.

graouts committed:
[web-animations] start time of animations created during a page rendering update should match that update's timeline time
https://bugs.webkit.org/show_bug.cgi?id=290993
rdar://138844796

Reviewed by Dean Jackson.

WebKit differs from Chrome and Firefox when setting the start time of animations
created during a page rendering update.

In WebKit, we mark any animation that is pending as ready when we run the "update
animations and send events" [0] step, which means that all animations pending at
this point will share the same start time and will also have a current time of 0
when the ready promise is resolved.

In Chrome and Firefox, animations created during a page rendering update will have
their start time set to the timeline time of that same update, even though their
ready promise may be resolved at a later time, for instance during the next page
rendering update. This means that the current time of animations will likely not
be 0 when the ready promise is resolved.

This is particularly visible for CSS-originated animations which may be created
during a page rendering update as styles are recalculated and layout is updated
a few steps after animations are updated. As such, WebKit typically is one rendering
frame behind Chrome and Firefox when dealing with CSS Transitions and CSS Animations.

This behavior is not actually a spec violation since the start time and time where
the ready promise is resolved is left as implementation-dependent [1].

However, this is proving to be a web compatibility issue since we found that captions
may flicker on YouTube for instance, and this is due to the behavior described above.
In fact, for CSS Transitions, we introduced some ad-hoc current time computation code
in 226359@main and 226577@main to deal with some compatibility issues (which we will
remove in a followup to this patch).

In light of this further web compatibility issue, we change how we determine the start
time of animations to align with Chrome and Firefox.

When we cache the current time during a page rendering update, we enqueue a task in
`AnimationTimelinesController::cacheCurrentTime()` to run once page rendering update
is done. Any animation that is created in this interval is added to a list of pending
animations, and when our queued that task is performed we set the "pending start time"
of those pending animations.

Then during the next page rendering update, as `WebAnimation::tick()` is called, we
check whether we have a "pending start time" and use that time to be the animation's
committed start time.

This now means that any animation created during a page rendering update, including
`requestAnimationFrame()` callbacks and code responding to animation events, will
have its pending start time set to the timeline time used for that page rendering
update. During the _next_ page rendering update, those animations will be marked as
"ready" and their start time will be set to the recorded pending start time.

If an animation was created any time after the page rendering update steps had
completed, for instance in a `setTimeout()` call within a `requestAnimationFrame()`
callback, that animation will have its start time set to match the next page rendering
update and will be marked as ready during the following page rendering update.

This behavior is now tested with an additional subtest in the existing WPT test
`web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html`.
While that behavior isn't technically normative, as noted above, it seems like important
enough for Web compatibility that it should be part of the WPT suite.

The subtle changes in timing required some adaptation of WebKit-specific tests. For the
compositing tests, because animations and transitions start a frame sooner, we had to update
those tests to have an animation transform in the output and to make sure the animations were
longer to yield a stable animated first frame in the output.

[0] https://drafts.csswg.org/web-animations-1/#update-animations-and-send-events
[1] https://drafts.csswg.org/web-animations-1/#waiting-for-the-associated-effect

* LayoutTests/compositing/backing/backface-visibility-flip-expected.txt:
* LayoutTests/compositing/backing/backface-visibility-flip.html:
* LayoutTests/compositing/backing/backing-store-attachment-empty-keyframe.html:
* LayoutTests/compositing/backing/transition-extent-expected.txt:
* LayoutTests/compositing/backing/transition-extent.html:
* LayoutTests/compositing/layer-creation/overlap-animation-clipping.html:
* LayoutTests/compositing/layer-creation/overlap-animation-container.html:
* LayoutTests/compositing/layer-creation/overlap-animation-expected.txt:
* LayoutTests/compositing/layer-creation/overlap-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/timing-model/animations/setting-the-start-time-of-an-animation.html:
* LayoutTests/platform/ios/compositing/backing/transition-extent-expected.txt:
* LayoutTests/platform/mac/compositing/visible-rect/animated-from-none-expected.txt:
* LayoutTests/webanimations/css-transition-retargeting-to-same-value-upon-completion-with-timeout.html:
* LayoutTests/webanimations/custom-effect/custom-effect.html:
* LayoutTests/webanimations/offset-anchor-animation-yields-compositing.html:
* LayoutTests/webanimations/offset-distance-animation-yields-compositing.html:
* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::animationTimingDidChange):
* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::detachFromDocument):
(WebCore::AnimationTimelinesController::addPendingAnimation):
(WebCore::AnimationTimelinesController::cacheCurrentTime):
(WebCore::AnimationTimelinesController::processPendingAnimations):
(WebCore::AnimationTimelinesController::maybeClearCachedCurrentTime): Deleted.
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::removeReplacedAnimations): Use the live current time instead
of the cached current time so that we can guarantee that "remove" events are dispatched
after "finish" events for the same animation.
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::runPendingPlayTask): use the "pending start time" if set.
(WebCore::WebAnimation::runPendingPauseTask): use the "pending start time" if set.
(WebCore::WebAnimation::tick):
(WebCore::WebAnimation::maybeMarkAsReady): move "ready" logic into a dedicated method.
* Source/WebCore/animation/WebAnimation.h:
(WebCore::WebAnimation::setPendingStartTime):

Canonical link: https://commits.webkit.org/294049@main
